pyopensprinkler/__init__.py

```diff
@@ -1,84 +1,213 @@
+        for i, _ in enumerate(self._state["stations"]["snames"]):
+            if i not in self._stations:
+                self._stations[i] = Station(self, i)
+
+    def _refresh_state(self):
+        (_, content) = self.request("/ja")
         self._state = content
 
+    def _get_option(self, option):
+        """Retrieve option"""
+        return self._state["options"][option]
+
+    def _get_options(self):
+        """Retrieve options"""
+        return self._state["options"]
+
+    def _set_option(self, option, value):
+        """Set option"""
+        params = {option: value}
+        (_, content) = self.request("/co", params)
+        return content["result"]
+
+    def _set_options(self, options):
+        """Set options"""
+        (_, content) = self.request("/co", options)
+        return content["result"]
+
+    def _get_variable(self, option):
+        """Retrieve variable"""
+        return self._state["settings"][option]
+
+    def _get_variables(self):
+        """Retrieve variables"""
+        return self._state["settings"]
+
+    def _set_variable(self, variable, value):
+        """Set variable"""
+        params = {variable: value}
+        (_, content) = self.request("/cv", params)
+        return content["result"]
+
+    def _set_variables(self, variables):
+        """Set variables"""
+        (_, content) = self.request("/cv", variables)
+        return content["result"]
+
+    def enable(self):
+        """Enable operation"""
+        return self._set_variable("en", 1)
+
+    def disable(self):
+        """Disable operation"""
+        return self._set_variable("en", 0)
+
+    @property
+    def firmware_version(self):
+        """Retrieve firmware version"""
+        return self._get_option("fwv")
+
+    @property
+    def hardware_version(self):
+        """Retrieve hardware version"""
+        return self._get_option("hwv")
+
+    # lrun [station index, program index, duration, end time]
+    @property
+    def last_run_station(self):
+        """Retrieve last run station"""
+        return self._get_variable("lrun")[0]
+
+    @property
+    def last_run_program(self):
+        """Retrieve last run program"""
+        return self._get_variable("lrun")[1]
+
+    @property
+    def last_run_duration(self):
+        """Retrieve last run duration"""
+        return self._get_variable("lrun")[2]
+
+    @property
+    def last_run_end_time(self):
+        """Retrieve last run end time"""
+        return self._get_variable("lrun")[3]
+
+    @property
+    def rain_delay_enabled(self):
+        """Retrieve rain delay enabled"""
+        return bool(self._get_variable("rd"))
+
+    @property
+    def rain_delay_stop_time(self):
+        """Retrieve rain delay stop time"""
+        return self._get_variable("rdst")
+
+    @property
+    def rain_sensor_enabled(self):
+        """Retrieve rain sensor enabled"""
+        try:
+            return bool(self._get_variable("rs"))
+        except KeyError:
+            return False
```

Comment:
I think maybe this should return `None` instead

Reply:
This is for the legacy stuff yeah? I’d be OK with that..

Reply:
Done.